no-unused-prop-types sometimes fails to detect props in functions on stateless components

Here is a minimal example that triggers this:

``` jsx
const PagingBlock = function(props) {
  return (
    <span>
      <a onClick={() => props.previousPage()}/>
      <a onClick={() => props.nextPage()}/>
    </span>
 );
};

PagingBlock.propTypes = {
  nextPage: React.PropTypes.func.isRequired,
  previousPage: React.PropTypes.func.isRequired,
};
```

The error message is:

```
  18:17  error  'previousPage' PropType is defined but prop is never used  react/no-unused-prop-types
```

Notes:
- Converting this to a React.createClass with static propTypes fixes the issue
- Removing either `<a/>` tag produces the correct lint error, the false error only occurs when both are present
- The error is notified for the first tag (switching `nextPage` and `previousPage` inside the render function will make the error text say `nextPage` is defined but not used

Probably same issue...

``` 
import React, { PropTypes } from 'react';

export default function Article(props) {
    const {
        homepageThumb: {
            url: imageUrl,
            width: imageWidth,
            height: imageHeight
        }
    } = props;

    return (

        <img
            src={imageUrl}
            width={imageWidth}
            height={imageHeight}
            alt="thumb"
        />

    );
}

Article.propTypes = {
    homepageThumb: PropTypes.shape({
        url: PropTypes.string,
        width: PropTypes.number,
        height: PropTypes.number
    }).isRequired
};
```
```
  59:20  error    'thumbnail_images.homepage-thumb.width' PropType is defined but prop is never used   react/no-unused-prop-types
  60:21  error    'thumbnail_images.homepage-thumb.height' PropType is defined but prop is never used  react/no-unused-prop-types
```

--- a/lib/util/Components.js
+++ b/lib/util/Components.js
@@ -362,12 +362,13 @@ function componentRule(rule, context) {
         var isFunction = /Function/.test(node.type); // Functions
         var isMethod = node.parent && node.parent.type === 'MethodDefinition'; // Classes methods
         var isArgument = node.parent && node.parent.type === 'CallExpression'; // Arguments (callback, etc.)
+        var isJSX = node.parent && node.parent.type === 'JSXExpressionContainer';
         // Stop moving up if we reach a class or an argument (like a callback)
         if (isClass || isArgument) {
           return null;
         }
         // Return the node if it is a function that is not a class method
-        if (isFunction && !isMethod) {
+        if (isFunction && !isMethod && !isJSX) {
           return node;
         }
         scope = scope.upper;

--- a/lib/util/Components.js
+++ b/lib/util/Components.js
@@ -362,13 +362,14 @@ function componentRule(rule, context) {
         var isFunction = /Function/.test(node.type); // Functions
         var isMethod = node.parent && node.parent.type === 'MethodDefinition'; // Classes methods
         var isArgument = node.parent && node.parent.type === 'CallExpression'; // Arguments (callback, etc.)
-        var isJSX = node.parent && node.parent.type === 'JSXExpressionContainer';
+        // Attribute Expressions inside JSX Elements (<button onClick={() => props.handleClick()}></button>)
+        var isJSXExpressionContainer = node.parent && node.parent.type === 'JSXExpressionContainer';
         // Stop moving up if we reach a class or an argument (like a callback)
         if (isClass || isArgument) {
           return null;
         }
-        // Return the node if it is a function that is not a class method
-        if (isFunction && !isMethod && !isJSX) {
+        // Return the node if it is a function that is not a class method and is not inside a JSX Element
+        if (isFunction && !isMethod && !isJSXExpressionContainer) {
           return node;
         }
         scope = scope.upper;

--- a/tests/lib/rules/no-unused-prop-types.js
+++ b/tests/lib/rules/no-unused-prop-types.js
@@ -1428,6 +1428,39 @@ ruleTester.run('no-unused-prop-types', rule, {
         '};'
       ].join('\n'),
       parserOptions: parserOptions
+    }, {
+      // `no-unused-prop-types` in jsx expressions - [Issue #885]
+      code: [
+        'const PagingBlock = function(props) {',
+        '  return (',
+        '    <span>',
+        '      <a onClick={() => props.previousPage()}/>',
+        '      <a onClick={() => props.nextPage()}/>',
+        '    </span>',
+        ' );',
+        '};',
+
+        'PagingBlock.propTypes = {',
+        '  nextPage: React.PropTypes.func.isRequired,',
+        '  previousPage: React.PropTypes.func.isRequired,',
+        '};'
+      ].join('\n'),
+      parserOptions: parserOptions
+    }, {
+      // `no-unused-prop-types` rest param props in jsx expressions - [Issue #885]
+      code: [
+        'const PagingBlock = function(props) {',
+        '  return (',
+        '    <SomeChild {...props} />',
+        ' );',
+        '};',
+
+        'PagingBlock.propTypes = {',
+        '  nextPage: React.PropTypes.func.isRequired,',
+        '  previousPage: React.PropTypes.func.isRequired,',
+        '};'
+      ].join('\n'),
+      parserOptions: parserOptions
     }
   ],
 

